File: sdk/packages/python/iii/src/iii/payload.py
"""Payload redaction + truncation for invocation event capture."""

from __future__ import annotations

import json
import os
from typing import Any, Optional

REDACTED_PLACEHOLDER = "[REDACTED]"
_TRUNCATION_MARKER = '..."[TRUNCATED]"'


def resolve_max_bytes_from_env() -> Optional[int]:
    raw = os.environ.get("III_TRACE_PAYLOAD_MAX_BYTES")
    if raw is None:
        return None
    trimmed = raw.strip()
    if not trimmed or trimmed.lower() == "unlimited":
        return None
    try:
        parsed = int(trimmed)
    except ValueError:
        return None
    if parsed <= 0:
        return None
    return parsed

_SENSITIVE_FRAGMENTS = (
    "api_key",
    "apikey",
    "api-key",
    "password",
    "secret",
    "credential",
    "authorization",
    "auth_token",
    "access_token",
    "refresh_token",
    "bearer",
    "private_key",
    "client_secret",
)


def _is_sensitive_key(key: str) -> bool:
    lower = key.lower()
    if any(fragment in lower for fragment in _SENSITIVE_FRAGMENTS):
        return True
    # ``token`` alone is too common a substring; require whole-key or suffix match.
    return lower == "token" or lower.endswith("_token") or lower.endswith("-token")


def redact(value: Any) -> Any:
    if isinstance(value, dict):
        return {
            k: REDACTED_PLACEHOLDER if _is_sensitive_key(k) else redact(v)
            for k, v in value.items()
        }
    if isinstance(value, list):
        return [redact(item) for item in value]
    if isinstance(value, tuple):
        return tuple(redact(item) for item in value)
    return value


def redact_and_truncate(
    value: Any, max_bytes: Optional[int] = None
) -> tuple[str, bool]:
    redacted = redact(value)
    try:
        serialized = json.dumps(redacted, default=str, ensure_ascii=False)
    except (TypeError, ValueError):
        serialized = "null"

    if max_bytes is None or max_bytes <= 0:
        return serialized, False

    encoded = serialized.encode("utf-8")
    if len(encoded) <= max_bytes:
        return serialized, False

    marker_len = len(_TRUNCATION_MARKER.encode("utf-8"))
    if max_bytes <= marker_len:
        return _TRUNCATION_MARKER[:max_bytes], True

    cap = max_bytes - marker_len
    # Walk back to a UTF-8 boundary so we don't emit half-codepoints.
    cut = cap
    while cut > 0 and (encoded[cut] & 0xC0) == 0x80:
        cut -= 1
    truncated = encoded[:cut].decode("utf-8", errors="ignore") + _TRUNCATION_MARKER
    return truncated, True

File: sdk/packages/node/iii/src/telemetry-system/payload.ts
/** Payload redaction + truncation for invocation event capture. */

export const REDACTED_PLACEHOLDER = '[REDACTED]'
const TRUNCATION_MARKER = '..."[TRUNCATED]"'

export function resolveMaxBytesFromEnv(): number | null {
  const raw = process.env.III_TRACE_PAYLOAD_MAX_BYTES
  if (raw === undefined) return null
  const trimmed = raw.trim()
  if (trimmed === '' || trimmed.toLowerCase() === 'unlimited') return null
  // parseInt accepts partial parses like "8192mb" → 8192; Python int()
  // and Rust parse::<usize>() reject those. Match the strict semantics.
  if (!/^\d+$/.test(trimmed)) return null
  const parsed = Number(trimmed)
  if (parsed <= 0) return null
  return parsed
}

const SENSITIVE_FRAGMENTS = [
  'api_key',
  'apikey',
  'api-key',
  'password',
  'secret',
  'credential',
  'authorization',
  'auth_token',
  'access_token',
  'refresh_token',
  'bearer',
  'private_key',
  'client_secret',
]

function isSensitiveKey(key: string): boolean {
  const lower = key.toLowerCase()
  if (SENSITIVE_FRAGMENTS.some((fragment) => lower.includes(fragment))) return true
  // `token` alone is too common a substring; require whole-key or suffix match.
  return lower === 'token' || lower.endsWith('_token') || lower.endsWith('-token')
}

/** Recursively redact values of sensitive keys. Returns a new value. */
export function redact(value: unknown): unknown {
  if (value === null || value === undefined) return value
  if (Array.isArray(value)) return value.map(redact)
  if (typeof value === 'object') {
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[k] = isSensitiveKey(k) ? REDACTED_PLACEHOLDER : redact(v)
    }
    return out
  }
  return value
}

/** Redact then serialize to JSON, optionally capped at `maxBytes`. */
export function redactAndTruncate(
  value: unknown,
  maxBytes: number | null = null,
): { json: string; truncated: boolean } {
  const redacted = redact(value)
  let serialized: string
  try {
    serialized = JSON.stringify(redacted) ?? 'null'
  } catch {
    serialized = 'null'
  }

  if (maxBytes === null || maxBytes === undefined || maxBytes <= 0) {
    return { json: serialized, truncated: false }
  }

  const byteLen = Buffer.byteLength(serialized, 'utf8')
  if (byteLen <= maxBytes) {
    return { json: serialized, truncated: false }
  }

  const markerLen = Buffer.byteLength(TRUNCATION_MARKER, 'utf8')
  // When maxBytes is smaller than the marker itself, emit only a
  // truncated marker so the result never exceeds the cap.
  if (maxBytes <= markerLen) {
    return { json: TRUNCATION_MARKER.slice(0, maxBytes), truncated: true }
  }

  const cap = maxBytes - markerLen
  const buf = Buffer.from(serialized, 'utf8')
  let cut = Math.min(cap, buf.length)
  // Walk back to a UTF-8 boundary so we don't emit half-codepoints.
  while (cut > 0 && (buf[cut] & 0xc0) === 0x80) {
    cut -= 1
  }
  const truncated = buf.subarray(0, cut).toString('utf8') + TRUNCATION_MARKER
  return { json: truncated, truncated: true }
}
